## Introduction
The Human Protein Atlas is a Swedish-based program initiated in 2003 with the aim to map all the human proteins in cells, tissues and organs using integration of various omics technologies, including antibody-based imaging, mass spectrometry-based proteomics, transcriptomics and systems biology. All the data in the knowledge resource is open access to allow scientists both in academia and industry to freely access the data for exploration of the human proteome. 
The Human Protein Atlas consists of six separate parts, each focusing on a particular aspect of the genome-wide analysis of the human proteins.
- The [Tissue Atlas](https://www.proteinatlas.org/humanproteome/tissue) showing the distribution of the proteins across all major tissues and organs in the human body
- The [Cell Atlas](https://www.proteinatlas.org/humanproteome/cell) showing the subcellular localization of proteins in single cells
- The [Pathology Atlas](https://www.proteinatlas.org/humanproteome/pathology) showing the impact of protein levels for survival of patients with cancer
- The [Blood Atlas](https://www.proteinatlas.org/humanproteome/blood)
- The [Brain Atlas](https://www.proteinatlas.org/humanproteome/brain)
- Tthe [Metabolic Atlas](https://www.proteinatlas.org/humanproteome/metabolic)

### The Tissue Atlas
The Tissue Atlas shows the expression and localization of human proteins across tissues and organs, based on deep sequencing of RNA (RNA-seq) from 37 major different normal tissue types and immunohistochemistry on tissue microarrays containing 44 different tissue types. Altogether 76 different cell types, corresponding to 44 normal human tissue types covering all major parts of the human body, have been analyzed manually and the data is presented as histology-based annotation of protein expression levels. The antibody-based protein profiles are qualitative and describe the spatial distribution, cell type specificity and the rough relative abundance of proteins in these tissues, whereas the mRNA data provide quantitative data on the average gene expression within an entire tissue. For each gene, the immunohistochemical staining profile is matched with mRNA data and gene/protein characterization data to yield an "annotated protein expression" profile.

### The Cell Atlas
The Cell Atlas provides high-resolution insights in the spatial distribution of proteins within cells. Firstly, it contains mRNA expression profiles from a diverse panel of human-derived cell lines (n=64) representing different germ layers and tissues. Secondly, the atlas contains high-resolution, multicolour immunofluorescence images of cells that detail the subcellular distribution pattern of proteins in these cells. By default, U-2OS cells and 2 other cell lines, selected based on expression, are probed with each antibody. The cells are stained in a standardized way where the antibody of interest is visualized in green, the microtubules red, the endoplasmic reticulum yellow, and nuclei counterstained in blue. The images are manually annotated in terms of spatial distribution to 32 different cellular structures representing 14 major organelles. The annotated locations for every protein are classified as main and additional, and assigned a reliability score.

### The Pathology Atlas
The Human Pathology Atlas is based on a systems-based analysis of the transcriptome of 17 main cancer types using data from 8000 patients. In addition, we show a new concept to present patient survival data, called Interactive Survival Scatter plots, and in the atlas, we present more than 400,000 plots. A national supercomputer center were used to analyze more than 2.5 petabytes of underlying publicly available data from the Cancer Genome Atlas (TCGA) to generate more than 900,000 survival plots describing the consequence of RNA and protein levels on clinical survival. The Pathology Atlas also contains 5 million pathology-based images generated by the Human Protein Atlas consortium. The research reports several important findings related to cancer biology and treatment. Firstly, many genes are differentially expressed in cancers, and a large proportion of these genes have an impact on overall patient survival. The research also showed that gene expression patterns of individual tumors varied considerably, and could exceed the variation observed between different cancer types. Shorter patient survival was generally associated with up-regulation of genes involved in mitosis and cell growth, and down-regulation of genes involved in cellular differentiation. The data allowed for generation of personalized genome-scale metabolic models for cancer patients to identify key genes involved in tumor growth.

### The Brain Atlas
The Brain Atlas explores the protein expression in the mammalian brain by integration of data from three mammalian species (human, pig and mouse). Transcriptomics data is combined with antibody-based protein localization in human samples and whole mouse brain. Protein-coding human genes (and one-to-one orthologues in pig and mouse) is provided with a brain-centric summary page, showing available expression data (mRNA) for brain samples grouped into 10 main brain regions, as well as data for pituitary gland, retina, corpus callosum and spinal cord. Series of sections, representing the whole mouse brain, analyzed for 271 genes are available as large 100 megapixel immunofluorescence images with microscopic resolution to explore the protein location in mouse brain (also summarized into 127 regions of interest). A selection of human localization data is also imported/linked from the Tissue Atlas and summarized on the Brain Atlas.

### The Blood Atlas
The Blood Atlas contributes with data regarding the cell types and proteome of human blood. Transcriptomics data from 18 single blood cell types isolated by cell sorting, including various B- and T-cells, monocytes, granulocytes and dendritic cells, is provided. The blood proteome is represented by data on concentration of blood proteins determined by mass spectrometry-based proteomics and/or antibody-based immune assays. In addition, a categorisation of the human secretome is presented, where the human predicted secreted proteins have been annotated in an attempt to determine which genes are actively secreted to human blood and which have more local functions in compartments such as the digestive system, brain or reproductive tissues.

### The Metabolic Atlas
The Tissue Atlas was expanded with content from Metabolic Atlas (metabolicatlas.org) to enable exploration of protein function and gene expression in the context of the human metabolic network. Manually curated maps are available for over 120 different metabolic pathways or subsystems, each depicting the association of proteins with the involved biochemical reactions. Below each pathway map, a heatmap shows the mRNA expression of all proteins associated with the pathway across 37 different tissue types. Further detail and fully interactive pathway maps are available on metabolicatlas.org, which provides additional information on thousands of network components such as reactions, metabolites, and genes.

## Download data
**Data from the Human Protein Atlas in XML format:**

The XML file contains most of the data in the Human Protein Atlas version 19.3, including **protein expression data** (in normal and tumor tissues and in cell lines), **antigen sequences**, **Western blot data for antibodies**, **protein array data for antibodies**, **RNA-seq data**, **external references such as UniProt identifiers**, and more. The data is based on Ensembl version 92.38. The file structure is presented in the XSD-schema. This data can also be downloaded for a resulting gene set when using the search function (via the xml link on the result page).
The XML file presented here is compressed with gzip due to its size. It can be uncompressed with an archive program like 7â€‘zip.

```
wget https://www.proteinatlas.org/download/proteinatlas.xml.gz
```

**proteinatlas xml file (partial)**

```
<?xml version="1.0" encoding="UTF-8"?>
<proteinAtlas xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://v19.proteinatlas.org/download/proteinatlas.xsd" schemaVersion="2.6">
	<entry version="19" url="http://v19.proteinatlas.org/ENSG00000000003">
		<name>TSPAN6</name>
		<synonym>T245</synonym>
		<synonym>TM4SF6</synonym>
		<synonym>TSPAN-6</synonym>
		<identifier id="ENSG00000000003" db="Ensembl" version="92.38">
			<xref id="O43657" db="Uniprot/SWISSPROT"/>
		</identifier>
		<proteinClasses>
			...
		</proteinClasses>
		<proteinEvidence evidence="Evidence at protein level">
			...
		</proteinEvidence>
		<tissueExpression source="HPA" technology="IHC" assayType="tissue">
			...
		</tissueExpression>
		<pathologyExpression source="HPA" technology="RNA" assayType="pathology">
			...
		</pathologyExpression>
		<cellExpression source="HPA" technology="ICC/IF">
			...
		</cellExpression>
		<rnaExpression source="HPA" technology="RNAseq" assayType="consensusTissue">
			...
		</rnaExpression>
		...
		<antibody id="HPA004109" releaseVersion="2" releaseDate="2006-10-30" RRID="AB_1080301">
			<antigenSequence source="HPA">NSFKNNYEKALKQYNSTGDYRSHAVDKIQNTLHCCGVTDYRDWTDTNYYSEKGFPKSCCKLEDCTPQRDADKVNNEG</antigenSequence>
			<antibodyTargetWeights source="HPA">
				<weight value="27.6" unit="kDa"/>
				<weight value="15.6" unit="kDa"/>
				<weight value="15.0" unit="kDa"/>
			</antibodyTargetWeights>
			...
			<cellExpression source="HPA" technology="ICC/IF">
				...
			</cellExpression>
			<proteinArray source="HPA" technology="PA">
				...
			</proteinArray>
		</antibody>
	</entry>
</proteinAtlas>
```


## Reference
1. https://www.proteinatlas.org/about/download
2. https://www.proteinatlas.org/about
