## 1. Introduction of Cell Ranger
Cell Ranger is a set of analysis pipelines that process Chromium single-cell RNA-seq output to align reads, generate feature-barcode matrices and perform clustering and gene expression analysis. Cell Ranger includes four pipelines relevant to single-cell gene expression experiments:
- **cellranger mkfastq** demultiplexes raw base call (BCL) files generated by Illumina sequencers into FASTQ files. It is a wrapper around Illumina's bcl2fastq, with additional useful features that are specific to 10x libraries and a simplified sample sheet format.
- **cellranger count** takes FASTQ files from cellranger mkfastq and performs alignment, **filtering, barcode counting, and UMI counting**. It uses the Chromium cellular barcodes to **generate feature-barcode matrices, determine clusters, and perform gene expression analysis**. The count pipeline can take input from multiple sequencing runs on the same GEM well. cellranger count also processes Feature Barcoding data alongside Gene Expression reads.
- **cellranger aggr** aggregates outputs from multiple runs of cellranger count, normalizing those runs to the same sequencing depth and then recomputing the feature-barcode matrices and analysis on the combined data. The aggr pipeline can be used to combine data from multiple samples into an experiment-wide feature-barcode matrix and analysis.
- **cellranger reanalyze** takes feature-barcode matrices produced by cellranger count or cellranger aggr and reruns the **dimensionality reduction, clustering, and gene expression** algorithms using tunable parameter settings.

These pipelines combine Chromium-specific algorithms with the widely used RNA-seq aligner STAR. Output is delivered in standard BAM, MEX, CSV, HDF5 and HTML formats that are augmented with cellular information.

## 2. System Requirements
### Hardware
Cell Ranger pipelines run on Linux systems that meet these minimum requirements:

- 8-core Intel or AMD processor (16 cores recommended)
- 64GB RAM (128GB recommended)
- 1TB free disk space
- 64-bit CentOS/RedHat 6.0 or Ubuntu 12.04

### Software
In order to run cellranger mkfastq, the following software needs to be installed:

**Illumina® bcl2fastq**: bcl2fastq must be version 2.17 or higher. It supports most sequencers running RTA version 1.18.54 or higher. If you are using NovaSeq™, the pipelines require version 2.20 or higher. If your sequencer is running an older version of RTA, then the pipelines require bcl2fastq 1.8.4.

## 3. Installing Cell Ranger
### Download and Extract Cell Ranger
1. Go to the Cell Ranger Download page, fill out the "10x Genomics End User Software License Agreement" information and copy the download command from the subsequent page. Paste the entire command onto your command line. Parts of the download command change periodically, so copying the same command seen here will not work. Be sure to copy the whole command from the Downloads page.
```
curl -o cellranger-3.1.0.tar.gz "http://cf.10xgenomics.com/releases/cell-exp/cellranger-3.1.0.tar.gz?Expires=1568276695&Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cDovL2NmLjEweGdlbm9taWNzLmNvbS9yZWxlYXNlcy9jZWxsLWV4cC9jZWxscmFuZ2VyLTMuMS4wLnRhci5neiIsIkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTU2ODI3NjY5NX19fV19&Signature=FLKeP1u95hziLa8ViQywYlipCeuTvdecksuVu~JIVsUcy7pev8VfwSwQ5kdst4hWB641KuD2qSM-5meVgCLOwE1h50oiEqYHlFfZ9XT30bAOb74cT5amOGX6HrX5nBGocLrH~gymixrkEU-z9v22TYGdHMAiqV84qvWZYKOUPZsrLR3umQuT2uF19xxbaXbUCKsVfrCSAY~0AVfEyyFYYiKnmANtEOZ1FNM0Lr6LoSq4a8o9wU9YyMYJCOfYm6JqP5QYnXUHxUMzMM4j3vJ-zASLuXito0RvrxYzgM~zNRRFwGjvr98o0xkKZ6EzGXmunA08aPlgRMG6J9t1b02NMg__&Key-Pair-Id=APKAI7S6A5RYOXBWRPDA"
```
2. Some systems do not have the command, curl (client url) installed. An alternative command is wget.
Note: You do not need to do this if the curl command above worked for you.

```
wget -O cellranger-3.1.0.tar.gz "http://cf.10xgenomics.com/releases/cell-exp/cellranger-3.1.0.tar.gz?Expires=1568198894&Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiaHR0cDovL2NmLjEweGdlbm9taWNzLmNvbS9yZWxlYXNlcy9jZWxsLWV4cC9jZWxscmFuZ2VyLTMuMS4wLnRhci5neiIsIkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTU2ODE5ODg5NH19fV19&Signature=MAsku-zG5j7hV9svst0pxw9WTJk3ilq3DmLb1obNHmjsOcIL8CF2T0UzYpiGQ2lTQraiPNQ2nzhLgH1liGP~7JjoZ~HousFx1w7POggIWYkhDu7ktlGQCVvjxBqZvfoWpgVvi1U4lpy7Ud2c0Hq1r3HphUH-ZiHqiGSKfTarUZEQOLLx~5CXz~moIM~TwAF0KDLwbAPkceylZUOTcdzBAH~IsD-LBXK-7WbBA5ZOFOXhAtrLWQnzCRU0HjnkXRUNtKGAJO7ApXkg8PcNQaEWBGiJ6hrnCsbPdjaVFhOcRd-6mtV-zg2WTt1kVcd8UjA4GAQ42mm1SIw5xgTkWc-~cw__&Key-Pair-Id=APKAI7S6A5RYOXBWRPDA"
```
3. Unpack the tarball using the tar command:

```
tar -zxvf cellranger-3.1.0.tar.gz
```
### Add Cell Ranger to Your $PATH
1. Cell Ranger is now installed. To run it requires entering the path to the executable command on the command line. We want to be able to run the Cell Ranger by simply entering the word cellranger on the command line. To accomplish this, we add it to your $PATH variable. Go into the cellranger-3.1.0 directory and enter pwd.

```
cd cellranger-3.1.0
pwd
```
The output looks something like this:

```
/mnt/home/user.name/yard/apps/cellranger-3.1.0 
```
2. Then use the export command to add it into the $PATH variable.

```
echo 'export PATH="/mnt/home/user.name/yard/apps/cellranger-3.0.2/cellranger":$PATH' >> ~/.bashrc
source ~/.bashrc
```
### Perform a Sitecheck
1. The purpose of sitecheck is to check your system to make sure it meets the system requirements for running the cellranger pipeline. Run the command and use the > symbol to direct the output to a file.

```
cellranger sitecheck > sitecheck.txt
```
2. Cellranger software support team can review your sitecheck results. Send it to us using cellranger upload from the command line.

```
cellranger upload your.email@email.com sitecheck.txt
```
### Perform a Testrun
Once the user limit for open files is adjusted, run cellranger testrun.

```
cellranger testrun --id=check_install
```
The output looks something like this:

```
/mnt/home/user.name/yard/apps/cellranger-3.1.0/cellranger-cs/3.1.0/bin
cellranger testrun (3.1.0)
Copyright (c) 2019 10x Genomics, Inc.  All rights reserved.
-------------------------------------------------------------------------------
 
Running Cell Ranger in test mode...
...
```
This process takes about five minutes to run and completes with the following statement:

```
Pipestance completed successfully!
2019-09-11 15:49:00 Shutting down.
Saving pipestance info to "check_install/check_install.mri.tgz"
```
This indicates a successful testrun and, by extension, a successful Cell Ranger install.


## 4. Running Pipelines
### Different Ways of Running Cell Ranger
There are three primary ways to run Cell Ranger:

1. **Single Server**: Cell Ranger can run directly on a dedicated server. This is the most straightforward approach and the easiest to troubleshoot.  
2. **Job Submission Mode**: Cell Ranger can run using a single node on the cluster. Less cluster coordination is required since all work is done on the same node. This method works well even with job schedulers that are not officially supported.
3. **Cluster Mode**: Cell Ranger can run using multiple nodes on the cluster. This method provides high performance, but is difficult to troubleshoot since cluster setup varies among institutions.

### Example Data Analysis
Exciting research is being done using the 10x Genomics Single Cell Gene Expression solution. This guide outlines how to perform the analysis, and what results 10x assays and software produce using data from a recent Nature publication “Single-cell transcriptomes of the regenerating intestine reveal a revival stem cell” (2019; doi: 10.1038/s41586-019-1154-y).

The Nature publication used an older version of Cell Ranger (2.0.0) for initial analysis. The raw data generated by Cell Ranger were loaded into third-party tools for secondary analysis. However, in this guide, the new version of Cell Ranger (3.1.0) and Loupe Cell Browser (3.1.1) is used to perform initial and secondary analysis. Using these tools with default settings, some major results reported in the Nature publication were reproduced.
#### About the Example
Intestinal tissue can repair itself after an injury such as irradiation. However, the molecular mechanisms that underlie the process are not fully understood. The LGR5+ crypt base columnar cells are thought to drive intestinal epithelium regeneration, but these cells are lost after injury, while regeneration still takes place.

To identify the cells responsible for the regeneration process, a group of researchers in Toronto used 10x Genomics Single Cell 3’ Gene Expression solution to profile mouse intestine cells, with and without irradiation.
#### Workflow Overview
This guide focuses on two samples, crypts (enriched from the whole epithelia) from normal and irradiated mice. See the figure below.
![example-workflow](https://github.com/fengxiaZhou/NCP-scRNA-seq/raw/master/images/example-workflow.png)
#### Perform the Data Analysis
Once the FASTQ files for each samples is generated, the data analysis begins. The **cellranger count** pipeline can perform read alignment, UMI counting, and secondary analysis (dimensionality reduction, clustering, and visualization) for a single sample.  
Given that these are mouse samples, the pre-built mouse reference is used. After determining the path to the reference folder and FASTQ files, run the cellranger count for the normal sample:

```
cd ./normal/
cellranger count --id=normal \
                 --transcriptome=/path/to/refdata-cellranger-mm10-3.0.0 \
                 --fastqs=./indepth_C05_MissingLibrary_1_HL5G3BBXX,./indepth_C05_MissingLibrary_1_HNNWNBBXX
```
#### Result file
After successfully completing the pipeline, find the outs/ directory for each run to review many useful result files. Start by checking the results using the following files:

- The **web_summary.html** contains summary metrics and automated secondary analysis results. 
- If the metrics look reasonable, load the cloupe.cloupe into **Loupe Cell Browser** to visualize and explore the results.
- For additional analysis, load the **filtered_feature_bc_matrix** into third-party tools, such as **Seurat**.
#### Compare Results from Different Samples

## 5. Understanding Output
### Summary View
The run summary can be viewed by clicking "Summary" in the top left corner. The summary metrics describe sequencing quality and various characteristics of the detected cells.

![web-summary](https://github.com/fengxiaZhou/NCP-scRNA-seq/raw/master/images/web-summary.png)

### Analysis View
The automated secondary analysis results can be viewed by clicking "Analysis" in the top left corner. The secondary analysis provides the following:

- A dimensional reduction analysis which projects the cells into a 2-D space (t-SNE)
- An automated clustering analysis which groups together cells that have similar expression profiles
- A list of genes that are differentially expressed between the selected clusters
- A plot showing the effect of decreased sequencing depth on observed library complexity
- A plot showing the effect of decreased sequencing depth on median genes per cell detected
![analysis-summary](https://github.com/fengxiaZhou/NCP-scRNA-seq/raw/master/images/analysis-summary.png)

### Feature Barcoding View
Cell Ranger 3.0 introduced Feature Barcoding analysis. If you have Cell Ranger with Feature Barcoding libraries, new metrics that are specific to the given Feature Barcoding libraries will appear in summary tab.

Below is an example web summary when the Antibody Capture library type is used:
![image](https://github.com/fengxiaZhou/NCP-scRNA-seq/raw/master/images/feature-antibody.png)

## Reference：
https://support.10xgenomics.com/single-cell-gene-expression/software/overview/welcome 

## 伪时间分析
伪时间分析建议采用 monocle3.0 软件
### 软件安装
```{r}
##安装依赖
BiocManager::install(c('BiocGenerics','DelayedArray', 'DelayedMatrixStats','limma', 'S4Vectors', 'SingleCellExperiment','SummarizedExperiment', 'batchelor'))
##安装monocle3
devtools::install_github('cole-trapnell-lab/leidenbase')
devtools::install_github('cole-trapnell-lab/monocle3')
library(monocle3)
```
### 标准分析流
```{r}
library(Seurat)
library(monocle3)
endo<-readRDS("../output/pbmc_tutorial.rds")
data <- endo@assays$RNA@counts
pd <-  endo@meta.data
```
不要直接把meta.data放入pd中去，太多没用的信息了，先清理下
```{r}
new_pd<-select(pd,nCount_RNA,nFeature_RNA,percent.mt)
new_pd$Cell.type<-Idents(endo)
head(new_pd)
                      nCount_RNA nFeature_RNA percent.mt
AAACCTGGTATCAGTC-1_1      4835         1383   3.536711
AACCGCGCATTCCTGC-1_1      4029         1230   2.655746
AACGTTGAGCCCAATT-1_1      2576          918   3.377329
AAGACCTGTGGTTTCA-1_1      2841         1044   5.455825
AAGTCTGCATGAAGTA-1_1      3731         1213   3.886358
ACAGCCGCATCGGACC-1_1      5915         1918   3.043111
                                     Cell.type
AAACCTGGTATCAGTC-1_1                       DC2
AACCGCGCATTCCTGC-1_1                      MDSC
AACGTTGAGCCCAATT-1_1                      MDSC
AAGACCTGTGGTTTCA-1_1 Dendritic.cells.activated
AAGTCTGCATGAAGTA-1_1                      MDSC
ACAGCCGCATCGGACC-1_1 Dendritic.cells.activated

fData <- data.frame(gene_short_name = row.names(data), row.names = row.names(data))
```
生成monocle的cds对象
```{r}
cds <- new_cell_data_set(data,cell_metadata  = new_pd,gene_metadata  = fData)
```
运行下游标准流程，无论如何都得跑，是必须的默认流程
```{r}
cds <- preprocess_cds(cds, num_dim = 30)
#umap
cds <- reduce_dimension(cds,umap.n_neighbors = 20L)
#color by seurat cluster
plot_cells(cds,label_groups_by_cluster=FALSE,color_cells_by = "Cell.type")
```
![markdown](https://github.com/ccgBiotechLover/single-cell-course/raw/master/monocle%20figures/fig10.webp "markdown")
图 10：undefined
```{r}
#cluster
cds <- cluster_cells(cds,resolution = 0.5)
#color by monocle cluster
plot_cells(cds, color_cells_by = "partition",label_groups_by_cluster=FALSE)
```
![markdown](https://github.com/ccgBiotechLover/single-cell-course/raw/master/monocle%20figures/fig11.webp "markdown")
图 11：undefined
```{r}
cds <- learn_graph(cds)
plot_cells(cds,color_cells_by = "Cell.type",label_groups_by_cluster=FALSE,label_leaves=TRUE,label_branch_points=TRUE)
```
![markdown](https://github.com/ccgBiotechLover/single-cell-course/raw/master/monocle%20figures/fig12.webp "markdown")
图 12：undefined
### 定义时间开始节点&&伪时间分析
```{r}
##一个有用的寻找起源节点的函数
get_earliest_principal_node <- function(cds, time_bin="Dendritic.cells.activated"){
  cell_ids <- which(colData(cds)[, "Cell.type"] == time_bin)
  closest_vertex <-
  cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
  closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])
  root_pr_nodes <-
  igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numeric(names
  (which.max(table(closest_vertex[cell_ids,]))))]
  root_pr_nodes
}

#order cell
cds <- order_cells(cds, root_pr_nodes=get_earliest_principal_node(cds))
#pseudotime analysis
plot_cells(cds,color_cells_by = "pseudotime",label_cell_groups=FALSE,label_leaves=FALSE,label_branch_points=FALSE,graph_label_size=1.5)
```
![markdown](https://github.com/ccgBiotechLover/single-cell-course/raw/master/monocle%20figures/fig13.webp "markdown"）
PS:对多样本进行伪时间分析，请参考[使用Monocle3对多样本单细胞数据进行伪时间分析](https://www.hongguangblog.cn/archives/48/ "普通链接带标题")
其他教程链接：
[整合刺激性和对照性PBMC数据集，以学习细胞类型特异性反应](https://www.hongguangblog.cn/archives/79/ "普通链接带标题")
[10X单细胞数据针对细胞及其亚型的基因集功能分析和转录调控分析](https://www.hongguangblog.cn/archives/13/ "普通链接带标题")
Reference:<https://www.hongguangblog.cn/archives/6/>
