## Generating FASTQs
Transform the bcl format to fastq  
```linux
cellranger mkfastq --id=tiny-bcl \
                     --run=/path/to/tiny_bcl \
                     --csv=cellranger-tiny-bcl-simple-1.2.0.csv
```
`--run`	(Required) The path of Illumina BCL run folder.  

`--id`	(Optional; defaults to the name of the flowcell referred to by --run) Name of the folder created by mkfastq.

`--csv`	(Optional) Path to a simple CSV with lane, sample, and index columns, which describe the way to demultiplex the flowcell. The index column should contain a 10x sample set name (e.g., SI-GA-A12 or the actual oligo sequence used). This is an alternative to the Illumina IEM sample sheet, and will be ignored if --samplesheet is specified.

##Specifying Input FASTQs
 If you ran `mkfastq`, your files will be in a (MKFASTQ_ID)/outs/fastq_path folder, and your file hierarchy probably looks something like this:
```
MKFASTQ_ID
|-- MAKE_FASTQS_CS
`-- outs
    |-- fastq_path
        |-- HFLC5BBXX
            |-- test_sample1
            |   |-- test_sample1_S1_L001_I1_001.fastq.gz
            |   |-- test_sample1_S1_L001_R1_001.fastq.gz
            |   |-- test_sample1_S1_L001_R2_001.fastq.gz
            |   |-- test_sample1_S1_L002_I1_001.fastq.gz
            |   |-- test_sample1_S1_L002_R1_001.fastq.gz
            |   |-- test_sample1_S1_L002_R2_001.fastq.gz
            |   |-- test_sample1_S1_L003_I1_001.fastq.gz
            |   |-- test_sample1_S1_L003_R1_001.fastq.gz
            |   `-- test_sample1_S1_L003_R2_001.fastq.gz
            |-- test_sample2
            |   |-- test_sample2_S2_L001_I1_001.fastq.gz
            |   |-- test_sample2_S2_L001_R1_001.fastq.gz
            |   |-- test_sample2_S2_L001_R2_001.fastq.gz
            |   |-- test_sample2_S2_L002_I1_001.fastq.gz
            |   |-- test_sample2_S2_L002_R1_001.fastq.gz
            |   |-- test_sample2_S2_L002_R2_001.fastq.gz
            |   |-- test_sample2_S2_L003_I1_001.fastq.gz
            |   |-- test_sample2_S2_L003_R1_001.fastq.gz
            |   `-- test_sample2_S2_L003_R2_001.fastq.gz
        |-- Reports
        |-- Stats
        |-- Undetermined_S0_L001_I1_001.fastq.gz
        ...
        `-- Undetermined_S0_L003_R2_001.fastq.gz
```

##V(D)J T, B Cell Analysis
To generate single-cell V(D)J sequences and annotations for a single library, run `cellranger vdj` with the following arguments. For a complete list of command-line arguments, run `cellranger vdj --help`.  
```
$ cellranger vdj --id=sample345 \
                 --reference=/opt/refdata-cellranger-vdj-GRCh38-alts-ensembl-3.1.0 \
                 --fastqs=/home/jdoe/runs/HAWT7ADXX/outs/fastq_path \
                 --sample=mysample
```
`--id`	A unique run ID string: e.g. sample345  

`--fastqs`	Path of the FASTQ folder generated by cellranger mkfastq
e.g. /home/jdoe/runs/HAWT7ADXX/outs/fastq_path  
Can take multiple comma-separated paths, which is helpful if the same library was sequenced on multiple flowcells.  
Doing this will treat all reads from the library, across flowcells, as one sample.  
If you have multiple libraries for the sample, you will need to run cellranger vdj using a custom MRO file as detailed on the Multi-Library Samples page.  

`--reference`	Path to the Cell Ranger V(D)J compatible reference e.g. `/opt/refdata-cellranger-vdj-GRCh38-alts-ensembl-3.1.0`. If `--denovo` is specified, this argument is optional.

`--sample`	Sample name as specified in the sample sheet supplied to `mkfastq`.
Can take multiple comma-separated values, which is helpful if the sample was sequenced on multiple flowcells and the sample name used (and therefore fastq file prefix) is not identical between them.
Doing this will treat all reads from the library, across flowcells, as one sample.
If you have multiple libraries for the sample, you will need to run `cellranger vdj` using a custom MRO file as detailed on the Multi-Library Samples page.
##Understanding Output
Output files will appear in the `outs/` subdirectory within this pipeline output directory.  
In this case,
    `/home/jdoe/runs/` is where the pipeline was run,
    `/home/jdoe/runs/sample345/` is the top-level output directory containing pipeline metadata, and
    `/home/jdoe/runs/sample345/outs/` contains the final pipeline output files.

The contents of this outs/ directory contain the data that is described in the remainder of this section:
- web_summary.html,
- metrics_summary.csv,
- annotation CSV/JSONs,
- FASTQ/FASTAs, and
- barcoded BAMs.


### Run Summary
The `cellranger vdj` pipeline outputs a summary HTML file containing summary metrics and automated secondary analysis results.  
#### Summary View
The run summary can be viewed by clicking "Summary" in the top left corner. The summary metrics describe sequencing quality and various characteristics of the detected cells.  
The number of cells detected, the mean read pairs per cell, and the number of V-J spanning productive paired cells are prominently displayed near the top of the page.

Click the '?' in the upper right corner of each dashboard for more information on each metric.

The Barcode Rank Plot under the "Cells" dashboard shows the count of filtered UMIs mapped to each barcode. A barcode must have a contig that aligns to a V segment to be identified as a targeted cell. (In the denovo case, the only requirement is a contig's presence.) There must also be at least three filtered UMIs with at least two read pairs each. It is possible that a barcode with at least as many filtered UMIs as another cell-associated barcode is not identified as a targeted cell. The color of the graph is based on the local density of cell-associated barcodes.

#### Analysis View  
The automated secondary analysis results can be viewed by clicking "Analysis" in the top left corner. The secondary analysis provides the following:

* A bar chart showing the the distribution of the top 10 clonotypes in this sample.
* A table of the top 10 clonotypes, their cell counts, their proportions, and the their associated sets of CDR3 amino acid sequences.

Note that because clonotypes are defined by productive sequences only, there are additional sequences that may not be displayed in the web summary. A more complete set of sequences associated with each cell can be found in the filtered contig files.

### Summary Metrics  
The `cellranger vdj` pipeline outputs `metrics_summary.csv` which contains a number of key metrics about the barcoding and sequencing process. Conventions for these metrics:

- In some cases there is a metric for each chain type. We only show one.
- Cells means targeted cells, i.e. T or B cells.
- References to quality scores exclude very low quality/no-call (Q ≤ 2) bases.  
```
Type 	Name                           	Description
int 	Estimated Number of Cells 	The number of barcodes estimated to correspond to GEMs containing cells. See Targeted Cell Calling Algorithm.
int 	Mean Read Pairs per Cell 	Number of input read pairs divided by the estimated number of cells.
int 	Number of Cells With Productive V-J Spanning Pair 	Number of cell barcodes for which at least one productive sequence was found for each of TRA and TRB (or heavy and light chains, for Ig).
int 	Number of Read Pairs 	Total number of read pairs that were assigned to this library in demultiplexing.
float 	Valid Barcodes 	Fraction of reads with barcodes that match the whitelist after barcode correction.
float 	Q30 Bases in Barcode 	Fraction of cell barcode bases with Q-score ≥ 30.
float 	Q30 Bases in RNA Read 1 	Fraction of read 1 bases with Q-score ≥ 30. (Likewise for read 2.)
float 	Q30 Bases in Sample Index 	Fraction of sample index bases with Q-score ≥ 30.
float 	Q30 Bases in UMI 	Fraction of UMI bases with Q-score ≥ 30.
float 	Reads Mapped to Any V(D)J Gene 	Fraction of reads that partially or wholly map to a V(D)J gene segment.
float 	Reads Mapped to TRA 	Fraction of reads that map partially or wholly to a TRA gene segment.
int 	Mean Used Read Pairs per Cell 	Mean number of read pairs used in assembly per cell barcode. These reads must have a cell barcode, map to a V(D)J gene, and have a UMI with sufficient read support, counted after subsampling.
float 	Fraction Reads in Cells 	Number of reads with cell barcodes divided by the number of reads with valid barcodes.
float 	Median TRA UMIs per Cell 	Median number of UMIs assigned to a TRA contig per cell.
float 	Paired Clonotype Diversity 	Effective diversity of the paired clonotypes, computed as the Inverse Simpson Index of the clonotype frequencies. A value of 1 indicates a minimally diverse sample - only one distinct clonotype was detected. A value equal to the estimated number of cells indicates a maximally diverse sample.
float 	Cells With TRA Contig 	Fraction of cell barcodes with at least one TRA contig annotated as a full or partial V(D)J gene.
float 	Cells With CDR3-annotated TRA Contig 	Fraction of cell barcodes with at least one TRA contig where a CDR3 was detected.
float 	Cells With V-J Spanning Contig 	Fraction of cell barcodes with at least one full-length contig.
float 	Cells With V-J Spanning TRA Contig 	Fraction of cell barcodes with at least one full-length TRA contig.
float 	Cells With Productive TRA Contig 	Fraction of cell barcodes with at least one full-length TRA contig that is productive.
float 	Cells With Productive V-J Spanning Pair 	Fraction of cell barcodes with at least one contig for each chain of the receptor pair that is productive.
```
### FASTA/FASTQ files  
The `cellranger vdj` pipeline outputs several indexed FASTA and FASTQ files.  

```
File type 	Primary Use Cases
FASTA 	Downstream tools such as the Integrated Genome Viewer (IGV) or V(D)J annotation tools like IGBLAST.
FASTQ 	Inspecting assembly base quality scores
```

```
File 	Records 	Description
filtered_contig.fasta 	Assembled contigs 	High-confidence contig sequences in cell barcodes.
filtered_contig.fastq 	Assembled contigs 	High-confidence contig sequences in cell barcodes.
all_contig.fasta 	Assembled contigs 	All assembled contig sequences.
all_contig.fastq 	Assembled contigs 	All assembled contig sequences.
consensus.fasta 	Clonotype consensus sequences 	Clonotype consensus sequences.
consensus.fastq 	Clonotype consensus sequences 	Clonotype consensus sequences.
concat_ref.fasta 	Concatenated germline segments 	Concatenated V(D)J germline segments for the segments detected on each consensus sequence. These serve as an approximate reference for each consensus sequence.
```
####Quality scores  
Typically, quality scores in a FASTQ file indicate the Phred-encoded probability that the base is correct. When a FASTQ file contains records for sequencing reads, the quality scores usually indicate the confidence of the base-caller at each base. Because `cellranger vdj` produces quality scores for assembled bases, the interpretation is slightly different.

```
File 	Interpretation
filtered_contig.fastq 	Probability that the base is not a sequencing, PCR, or reverse-transcription (RT) error. The quality score is computed using the per-read sequencing Q-scores and an assumed RT error rate.
all_contig.fastq 	Same as above.
consensus.fastq 	Same as above, except the reads used to assemble all of the contigs used to produce the consensus are used.
```
###V(D)J Annotations  
`cellranger vdj` pipeline produces V(D)J annotations on the assembled contigs and on the clonotype consensus sequences in multiple formats.  
####File type overview  
```
File type 	Description
CSV 	High-level annotations with one contig, consensus, or clonotype per row.
JSON 	Detailed annotations, including alignment coordinates and amino acid translations.
BED 	Germline V(D)J segments as features, for use with a tool like IGV.
```

####Annotation files  
```
File 	Description
clonotypes.csv 	High-level descriptions of each clonotype.
consensus_annotations.{csv,json} 	High-level and detailed annotations of each clonotype consensus sequence.
filtered_contig_annotations.csv 	High-level annotations of each high-confidence, cellular contig. This is a subset of all_contig_annotations.csv.
all_contig_annotations.{csv,bed,json} 	High-level and detailed annotations of each contig.
```
####Clonotype CSV file (clonotypes.csv)
```
Column 	Description
clonotype_id 	The ID of the clonotype to which this consensus sequence was assigned.
frequency 	The observed number of cell barcodes with this clonotype.
proportion 	The observed fraction of cell barcodes with this clonotype.
cdr3s_aa 	A semicolon-delimited list of chain:sequence pairs, where "chain" is e.g., TRA, TRB, IGK, IGL, or IGH and "sequence" is the CDR3 amino acid sequence for that chain.
cdr3s_nt 	A semicolon-delimited list of chain:sequence pairs, where "chain" is e.g., TRA, TRB, IGK, IGL, or IGH and "sequence" is the CDR3 nucleotide sequence for that chain.
```
####Contig annotation CSV files (*contig_annotations.csv)  
```
Column 	Description
barcode 	Cell-barcode for this contig.
is_cell 	True or False value indicating whether the barcode was called as a cell.
contig_id 	Unique identifier for this contig.
high_confidence 	True or False value indicating whether the contig was called as high-confidence (unlikely to be a chimeric sequence or some other artifact).
length 	The contig sequence length in nucleotides.
chain 	The chain associated with this contig; e.g., "TRA", "TRB", "IGK", "IGL", or "IGH". A value of "Multi" indicates that segments from multiple chains were present.
v_gene 	The highest-scoring V segment, e.g., TRAV1-1.
d_gene 	The highest-scoring D segment, e.g., TRBD1.
j_gene 	The highest-scoring J segment, e.g., TRAJ1-1.
c_gene 	The highest-scoring C segment, e.g., TRAC.
full_length 	If the contig was declared as full-length.
productive 	If the contig was declared as productive.
cdr3 	The predicted CDR3 amino acid sequence.
cdr3_nt 	The predicted CDR3 nucleotide sequence.
reads 	The number of reads aligned to this contig.
umis 	The number of distinct UMIs aligned to this contig.
raw_clonotype_id 	The ID of the clonotype to which this cell barcode was assigned.
raw_consensus_id 	The ID of the consensus sequence to which this contig was assigned.
```
####Consensus annotation CSV files (consensus_annotations.csv)
```
Column 	Description
clonotype_id 	The ID of the clonotype to which this consensus sequence was assigned.
consensus_id 	The ID of this consensus sequence.
```
###Barcoded BAMs
The `cellranger vdj` pipeline outputs several indexed BAM files. These files are primarily provided for use with a BAM visualization tool such as the Integrated Genome Viewer (IGV).
```
File 	Records 	Reference 	Description
all_contig.bam 	Reads 	Assembled contigs 	Reads aligned to assembled contigs, per cell barcode. This file demonstrates how the reads and UMIs support the assembled contigs within a cell barcode. Reads are not aligned across cell barcode boundaries. Please note that this BAM excludes reads whose barcodes don't match the whitelist, so it is not suitable as an archive of every single input read.
consensus.bam 	Contigs 	Clonotype consensus 	Each "reference" sequence is a clonotype consensus sequence, and each record is an alignment of a single cell's contig against this consensus. This file shows, for a clonotype consensus sequences, how the constituent per-cell assemblies support the consensus.
concat_ref.bam 	Contigs 	Concatenated germline segments 	Each reference sequence is, for each clonotype consensus, the annotated germline segments concatenated together. This file shows how both the per-cell contigs and the clonotype consensus contig relate to the germline reference. This file is expected to reveal polymorphisms, somatic mutations, and recombination-induced differences such as non-templated nucleotide additions.
```
The following assumes basic familiarity with the BAM format. More details on the SAM/BAM standard are available online.  
####BAM Barcode Tags  
Chromium cell barcode and UMI information for each read is stored as `TAG` fields:
```
Tag	Type	Description
CB	Z	Chromium cell barcode sequence that is error-corrected and confirmed against a list of known-good barcode sequences.
CR	Z	Chromium cell barcode sequence as reported by the sequencer.
CY	Z	Chromium cell barcode read quality. Phred scores as reported by sequencer.
UB	Z	Chromium UMI sequence that is error-corrected among other UMIs with the same cell barcode and gene alignment.
UR	Z	Chromium UMI sequence as reported by the sequencer.
UY	Z	Chromium UMI read quality. Phred scores as reported by sequencer.
BC	Z	Sample index read.
QT	Z	Sample index read quality. Phred scores as reported by sequencer.
```
The cell barcode `CB` tag includes a suffix with a dash separator followed by a number:
`AGAATGGTCTGCAT-1`

This number denotes what we call a GEM group, and is used to virtualize barcodes in order to achieve a higher effective barcode diversity when combining samples generated from separate GEM chip channel runs. Normally, this number will be "1" across all barcodes when analyzing a sample generated from a single GEM chip channel. It can either be left in place and treated as part of a unique barcode identifier, or explicitly parsed out to leave only the barcode sequence itself.  
####BAM CIGAR String  
The `cellranger vdj` pipeline uses the = and X CIGAR string operations to indicate matches and mismatches, respectively. This contrasts with most aligners which simply report M for match/mismatch. The SAM/BAM standard supports both CIGAR formats. For more details please refer to the SAM/BAM standard.

##V(D)J Secondary Analysis
###Comparing clonotype frequencies
The most high-level file that describes the clonotype frequencies for a sample is `clonotypes.csv`. See Annotations for details on this file. In order to compare clonotype frequencies between samples, you can match clonotypes using the `cdrs_nt` column, which contains the paired CDR3 nucleotide sequences for each clonotype.

###Inspecting consensus sequences  
Clonotype consensus sequences can be obtained from the `consensus.fastq` file, which also contains per-base quality scores.

Reference: <https://support.10xgenomics.com/single-cell-vdj/software/pipelines/latest/analysis>
