## Introduction
TCR-dist is an analytical tools to characterize these epitope-specific
repertoires: a distance measure on the space of TCRs that permits
clustering and visualization, a robust repertoire diversity metric
that accommodates the low number of paired public receptors
observed when compared to single-chain analyses, and a distancebased classifier that can assign previously unobserved TCRs to
characterized repertoires with robust sensitivity and specificity. 
## Requirements
**Python version**

- tested with 2.7.11

**Package dependencies**

- scipy: https://www.scipy.org/install.html tested with version 0.16.0

- scikit-learn: aka "sklearn" for KernelPCA, adjusted_mutual_info_score http://scikit-learn.org/stable/install.html tested with version 0.17

- matplotlib: tested with version 1.4.3

numpy: tested with version 1.10.1

**External command line tools**

- convert: (or rsvg-convert or inkscape) from Imagemagick is used to convert svg files to png files. If you have an alternative you can modify the function "convert_svg_to_png" in basic.py

- wget: (or curl) for downloading database and other files If you have something else that works similarly on your system, feel free to modify setup.py or contact me to add that as an option.

**NCBI Blast**

- The setup script will automatically download a compatible version into the ./external/ directory.

## Installation

```
## clone tcr-dist repository
git clone https://github.com/phbradley/tcr-dist.git
## Go to the tcr-dist/ directory (main code directory),run the command
## Note: If the download urls is not avaiable when downloading the BLAST, modify the source code of setup.py, change the available urls 
python setup.py

```
## Usage
The basic workflow starting from a sequence file would be to run

```
python run_basic_analysis.py --organism <organism> --pair_seqs_file <filename>
```
where <organism> is either mouse or human, and <filename> is the name of a .tsv (tab-separated values) file with the following fields:

```
id	epitope	subject	a_nucseq	b_nucseq	a_quals	b_quals
```
**id**: Unique identifier for the TCR

**epitope**: The name of the epitope to which the TCR binds

**subject**: The individual from whom the TCR was sampled, necessary for identifying clones, and analyzing repertoire privacy.

**a_nucseq, b_nucseq**: The nucleotide sequences of the TCR alpha and beta chain reads

**a_quals, b_quals**: Read quality information in the form of '.'-separated lists of the quality scores for the corresponding nucleotide sequences (a_nucseq and b_nucseq), used for read quality filtering (e.g., 25.32.35.36.45.45.36)

The id, a_quals, and b_quals fields can be omitted; see the help message printed by running the command:

```
python run_basic_analysis.py -h
```

This script will run a pipeline of analysis tools starting from three possible input filetypes:
```
- pair_seqs_file: A .tsv (tab-separated values) file with info on alpha and beta chain
       sequences and quality scores, epitope, subject, and id.
       Required fields: id epitope subject a_nucseq b_nucseq a_quals b_quals
       For further details, run "python read_pair_seqs.py -h"
       In particular,
          * if you don't have quality score info you can add --make_fake_quals and fictitious
            quality scores will be created.
          * if you don't have ids, you can add --make_fake_ids and default ids will be created based
            on position in the file.


- parsed_seqs_file: A processed sequence file with V and J genes assigned, CDR3s parsed, etc. Will be produced
       as an intermediate when you start from a pair_seqs_file and could be then reused to re-run downstream
       analyses. Could also be generated from the output of other tools (eg MIXCR, conversion scripts yet to come).

- clones_file: A processed version of the parsed_seqs_file in which per-TCR probabilities have been assigned and
       clones identified. One line per clone. Most of the scripts in the pipeline take a clones_file as input.
```
Use the corresponding command line option to point to the type of file you have.
    
The results of the analysis are summarized in an html file and associated directory of images which will be generated by the script. The default name (which can be changed with the --webdir option) will look something like 

```
<filename>*web/index.html
```

## Testing
In the tcr-dist/ directory you can run the command

```
python create_and_run_testing_datasets.py
```
which will create a directory called testing/, make two small dataset files there, and run the analysis pipeline on them.

You can also re-run the analysis in the original paper by typing:

```
python rerun_paper_analysis.py
```


## Reference
1. Article: [Quantifiable predictive features define epitope-specific T cell receptor repertoires](https://www.nature.com/articles/nature22383)
2. Github: https://github.com/phbradley/tcr-dist
